relayer/shannon/mode_trusted.go

```diff
@@ -0,0 +1,148 @@
+// The supplied account client is required to instantiate relay request signers corresponding to the apps held by the Trusted operation mode.
+func NewTrustedMode(appsPrivateKeys []*secp256k1.PrivKey, accountClient sdk.AccountClient) (*TrustedMode, error) {
+	appsSigners := make(map[string]signer)
+	for _, appPrivateKey := range appsPrivateKeys {
+		appAddr, err := getAddressFromPrivateKey(appPrivateKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO_TECHDEBT: update this code to drop the extra conversion to Hex once the shannon-sdk is udated to accept either a specific
+		// private key type or the generic PrivKey from github.com/cosmos/cosmos-sdk/crypto/types.
+		privateKeyHex := hex.EncodeToString(appPrivateKey.Bytes())
+		signer := signer{
+			accountClient: accountClient,
+			privateKeyHex: privateKeyHex,
+		}
+
+		appsSigners[appAddr] = signer
+	}
+
+	return &TrustedMode{
+		appsSigners: appsSigners,
+	}, nil
+}
+
+// TrustedMode represent an operation mode which behaves as follows:
+// 1. PATH (or more speicifcally the Shannon protocol integration instance) holds the private keys of users' app(s).
+// 2. Each relay request is signed by and sent on behalf of one of the held apps.
+// 3. Users need to select a specific app for each relay request, done using HTTP request's headers as of now.
+//
+// See the following link for more details on PATH's Trusted operation mode.
+// https://www.notion.so/buildwithgrove/Different-Modes-of-Operation-PATH-LocalNet-Discussions-122a36edfff6805e9090c9a14f72f3b5?pvs=4#122a36edfff680eea2fbd46c7696d845
+type TrustedMode struct {
+	// appsSigners is a map of application address to the corresponding signer, i.e. the signer that holds the app's private key.
+	// application address is used as the key to allow quick lookup of an app's corresponding signer for signing a relay request.
+	appsSigners map[string]signer
+}
+
+// BuildInstanceFromParams builds and returns a trustedModeInstance based on the supplied params map, which as of now is the HTTP request's headers.
+// It implements the OperationMode interface consumed by the Protocol struct.
+func (tm *TrustedMode) BuildInstanceFromHTTPRequest(httpReq *http.Request) (OperationModeInstance, error) {
+	if httpReq == nil || len(httpReq.Header) == 0 {
+		return nil, fmt.Errorf("TrustedMode BuildInstanceFromHTTPRequest: no HTTP headers supplied.")
+	}
+
+	selectedAppAddr := httpReq.Header.Get(headerAppAddr)
+	if selectedAppAddr == "" {
+		return nil, fmt.Errorf("TrustedMode BuildInstanceFromHTTPRequest: a target app must be supplied as HTTP header %s", headerAppAddr)
+	}
+
+	// TODO_DISCUSS: need a solid method of verifying the user that sent the HTTP request has access to the requested app's private key.
+	// This is needed to ensure only the app's owner can request that the relay be signed with the app's private key.
+	selectedAppSigner, ok := tm.appsSigners[selectedAppAddr]
```

Comment:
The `Gateway` should fetch the `Application`s delegation and check if the `Gateway` address is present.

Reply:
Fixed, thank you.